<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Classifier &mdash; bioy  documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="bioy  documentation" href="index.html" />
    <link rel="prev" title="Welcome to the bioy’s documentation page" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to the bioy’s documentation page"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">bioy  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-bioy_pkg.subcommands.classifier">
<span id="classifier"></span><h1>Classifier<a class="headerlink" href="#module-bioy_pkg.subcommands.classifier" title="Permalink to this headline">¶</a></h1>
<p>Classify sequences by grouping blast output by matching taxonomic names</p>
<p>Optional grouping by specimen and query sequences</p>
<div class="section" id="running-the-program">
<h2>Running the program<a class="headerlink" href="#running-the-program" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre>positional arguments:
  blast_file            CSV tabular blast file of query and subject hits.
  seq_info              File mapping reference seq name to tax_id
  taxonomy              Table defining the taxonomy for each tax_id

optional arguments:
  -h, --help            show this help message and exit
  --threads NUM         Number of threads (CPUs). Can also specify with
                        environment variable THREADS_ALLOC. [32]
  --copy-numbers CSV    Estimated 16s rRNA gene copy number for each
                        tax_ids (CSV file with columns: tax_id, median)
  --rank-thresholds CSV
                        Columns [tax_rank,tax_id,low,rank]
  --specimen-map CSV    CSV file with columns (name, specimen) assigning
                        sequences to groups. The default behavior is to
                        treat all query sequences as
                        belonging to one specimen.
  -w CSV, --weights CSV
                        Optional headless csv file with columns &#39;seqname&#39;,
                        &#39;count&#39; providing weights for each query sequence
                        described in the blast input (used, for example, to
                        describe cluster sizes for corresponding cluster
                        centroids).
  -o FILE, --out FILE   Classification results.
  -O FILE, --details-out FILE
                        Optional details of taxonomic assignments.
  --details-full        do not limit out_details to only larget cluster per
                        assignment
  --group-def INTEGER   define a group threshold for a particular rank
                        overriding --target-max-group-size. example:
                        genus:2 (NOT IMPLEMENTED)
  --has-header          specify this if blast data has a header
  --min-identity PERCENT
                        minimum identity threshold for accepting matches
                        [&gt;= 0.0]
  --max-identity PERCENT
                        maximum identity threshold for accepting matches
                        [&lt;= 100.0]
  --min-cluster-size INTEGER
                        minimum cluster size to include in classification
                        output [1]
  --min-coverage PERCENT
                        percent of alignment coverage of blast result [0.0]
  --specimen LABEL      Single group label for reads
  --starred PERCENT     Names of organisms for which at least one reference
                        sequence has pairwise identity with a query
                        sequence of at least PERCENT will be marked with an
                        asterisk[100.0]
  --target-max-group-size INTEGER
                        group multiple target-rank assignments that excede
                        a threshold to a higher rank [3]
  --target-rank TARGET_RANK
                        Rank at which to classify. Default: &quot;species&quot;
</pre></div>
</div>
<div class="section" id="positional-arguments">
<h3>Positional arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this headline">¶</a></h3>
<div class="section" id="blast-file">
<h4>blast_file<a class="headerlink" href="#blast-file" title="Permalink to this headline">¶</a></h4>
<p>A csv file with columns <strong>qseqid</strong>, <strong>sseqid</strong>, <strong>pident</strong>,
<strong>qstart</strong>, <strong>qend</strong> and <strong>qlen</strong>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The actual header is optional and if
present make sure to use the &#8211;has-header switch</p>
</div>
</div>
<div class="section" id="seq-info">
<h4>seq_info<a class="headerlink" href="#seq-info" title="Permalink to this headline">¶</a></h4>
<p>A csv file with minimum columns <strong>seqname</strong> and <strong>tax_id</strong>.  Additional
columns will be included in the details output.</p>
</div>
<div class="section" id="taxonomy">
<h4>taxonomy<a class="headerlink" href="#taxonomy" title="Permalink to this headline">¶</a></h4>
<p>A csv file with columns <strong>tax_id</strong>, <strong>rank</strong> and <strong>tax_name</strong>, plus at least
one additional rank column(s) creating a taxonomic tree such as <strong>species</strong>,
<strong>genus</strong>, <strong>family</strong>, <strong>class</strong>, <strong>pylum</strong>, <strong>kingdom</strong> and/or <strong>root</strong>.
The rank columns also give an order of specificity from right to left,
least specific to most specific respectively.</p>
</div>
</div>
<div class="section" id="optional-input">
<h3>Optional input<a class="headerlink" href="#optional-input" title="Permalink to this headline">¶</a></h3>
<div class="section" id="rank-thresholds">
<h4>rank-thresholds<a class="headerlink" href="#rank-thresholds" title="Permalink to this headline">¶</a></h4>
<p>This table defines the similarity thresholds at rank. The structure is
as follows:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="30%" />
<col width="15%" />
<col width="55%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">tax_id</th>
<th class="head">low</th>
<th class="head">target_rank</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>99</td>
<td>species</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>97</td>
<td>genus</td>
</tr>
<tr class="row-even"><td>1</td>
<td>95</td>
<td>family</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>The <strong>tax_id</strong> column identifies the subtree of the taxonomy to which
the threshold defined in <strong>low</strong> should be applied. The example
above is preloaded into the Classifier.  The threshold of 99 applies to all
tax_ids with an ancestor of tax_id=1 (ie, the entire taxonomy).</p>
<p>A custom example that can be specified by the user on top of the above example
more useful example might be this:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="30%" />
<col width="15%" />
<col width="55%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">tax_id</th>
<th class="head">low</th>
<th class="head">target_rank</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>2049</td>
<td>97</td>
<td>species</td>
</tr>
<tr class="row-odd"><td>2049</td>
<td>95</td>
<td>genus</td>
</tr>
<tr class="row-even"><td>2049</td>
<td>93</td>
<td>family</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>Here, all organisms belonging to family Actinomycetaceae (tax_id 2049)
will be classified to the species level using a threshold of 97%
instead of 99% in the first example, because we know that there is
more specific, species-level heterogeneity within this family.</p>
</div>
<div class="section" id="copy-numbers">
<h4>copy-numbers<a class="headerlink" href="#copy-numbers" title="Permalink to this headline">¶</a></h4>
<p>Below is an <em>example</em> copy numbers csv with the required columns:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="63%" />
<col width="19%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">tax_id</th>
<th class="head">tax_name</th>
<th class="head">median</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>155977</td>
<td>Acaryochloris</td>
<td>2.00</td>
</tr>
<tr class="row-odd"><td>155978</td>
<td>Acaryochloris marina</td>
<td>2.00</td>
</tr>
<tr class="row-even"><td>434</td>
<td>Acetobacter</td>
<td>5.00</td>
</tr>
<tr class="row-odd"><td>433</td>
<td>Acetobacteraceae</td>
<td>3.60</td>
</tr>
</tbody>
</table>
</div></blockquote>
</div>
<div class="section" id="weights">
<h4>weights<a class="headerlink" href="#weights" title="Permalink to this headline">¶</a></h4>
<p>Headerless file containing two columns specifying the seqname (clustername) and
weight (or number of sequences in the cluster).</p>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<div class="section" id="out">
<h4>out<a class="headerlink" href="#out" title="Permalink to this headline">¶</a></h4>
<p>A csv with columns and headers as in the example below:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="23%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">specimen</th>
<th class="head">assignment_id</th>
<th class="head">assignment</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>039_3</td>
<td>0</td>
<td>Pseudomonas mendocina;Pseudonocardia</td>
</tr>
<tr class="row-odd"><td>039_3</td>
<td>1</td>
<td>Rhizobiales</td>
</tr>
<tr class="row-even"><td>039_3</td>
<td>2</td>
<td>Alcaligenes faecalis*</td>
</tr>
<tr class="row-odd"><td>039_3</td>
<td>3</td>
<td>[no blast result]</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="21%" />
<col width="39%" />
<col width="39%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">low</th>
<th class="head">max_percent</th>
<th class="head">min_percent</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>95.00</td>
<td>99.02</td>
<td>95.74</td>
</tr>
<tr class="row-odd"><td>95.00</td>
<td>98.91</td>
<td>95.31</td>
</tr>
<tr class="row-even"><td>99.00</td>
<td>100.00</td>
<td>99.00</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="31%" />
<col width="17%" />
<col width="26%" />
<col width="26%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">target_rank</th>
<th class="head">reads</th>
<th class="head">pct_reads</th>
<th class="head">clusters</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>species</td>
<td>6</td>
<td>35.29</td>
<td>1</td>
</tr>
<tr class="row-odd"><td>genus</td>
<td>5</td>
<td>29.41</td>
<td>1</td>
</tr>
<tr class="row-even"><td>species</td>
<td>5
1</td>
<td>29.41
5.88</td>
<td>1
1</td>
</tr>
</tbody>
</table>
</div></blockquote>
</div>
<div class="section" id="details-out">
<h4>details-out<a class="headerlink" href="#details-out" title="Permalink to this headline">¶</a></h4>
<p>A csv that is basically a blast results breakdown of the <a class="reference internal" href="#out">out</a> output.</p>
</div>
</div>
</div>
<div class="section" id="internal-functions">
<h2>Internal functions<a class="headerlink" href="#internal-functions" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.agg_columns">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">agg_columns</tt><big>(</big><em>df</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#agg_columns"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.agg_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Create aggregate columns for assignments.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.assign">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">assign</tt><big>(</big><em>df</em>, <em>tax_dict</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#assign"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.assign" title="Permalink to this definition">¶</a></dt>
<dd><p>Create str assignment based on tax_ids str and starred boolean.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.assignment_id">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">assignment_id</tt><big>(</big><em>df</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#assignment_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.assignment_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets and drops the current dataframe&#8217;s
index and sets it to the assignment_hash</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.condense_ids">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">condense_ids</tt><big>(</big><em>df</em>, <em>tax_dict</em>, <em>ranks</em>, <em>max_group_size</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#condense_ids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.condense_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Create mapping from tax_id to its
condensed id and set assignment hash.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.focus_results_by_specificity">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">focus_results_by_specificity</tt><big>(</big><em>df</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#focus_results_by_specificity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.focus_results_by_specificity" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove less specific target_rank hits.</p>
<p>If a hit&#8217;s target rank shares taxonomy with a more specific
target hit then it will be dropped in favor
of the more specific target hit(s).</p>
<p>Example: Given a group of blast hits (by qseqid), if a hit matches a
Helicobacter pylori species sequence and also matches
a Helicobacter genus sequence the H. pylori sequence will be accepted
while the genus level hit will be dropped from the
group.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.freq">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">freq</tt><big>(</big><em>df</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#freq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.freq" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate pct_reads column.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.load_rank_thresholds">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">load_rank_thresholds</tt><big>(</big><em>path='/home/local/AMC/crosenth/src/bioy/bioy_pkg/data/rank_threshold_defaults.csv'</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#load_rank_thresholds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.load_rank_thresholds" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a rank-thresholds file.  If no argument is specified the default
rank_threshold_defaults.csv file will be loaded.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.most_specific">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">most_specific</tt><big>(</big><em>series</em>, <em>cols</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#most_specific"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.most_specific" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the most taxonomic specific tax_id available for the given
Series.</p>
<p>For most given Series the given target_rank is available and returned.
If the target_rank is NAN then return a less specific tax_id</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.pct_corrected">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">pct_corrected</tt><big>(</big><em>df</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#pct_corrected"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.pct_corrected" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate pct_corrected values.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.read_csv">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">read_csv</tt><big>(</big><em>filename</em>, <em>compression=None</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#read_csv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.read_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a csv file using pandas.read_csv with compression defined by
the file suffix unless provided.</p>
</dd></dl>

<dl class="function">
<dt id="bioy_pkg.subcommands.classifier.star">
<tt class="descclassname">bioy_pkg.subcommands.classifier.</tt><tt class="descname">star</tt><big>(</big><em>df</em>, <em>starred</em><big>)</big><a class="reference internal" href="_modules/bioy_pkg/subcommands/classifier.html#star"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#bioy_pkg.subcommands.classifier.star" title="Permalink to this definition">¶</a></dt>
<dd><p>Assign boolean if any items in the
dataframe are above the star threshold.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Classifier</a><ul>
<li><a class="reference internal" href="#running-the-program">Running the program</a><ul>
<li><a class="reference internal" href="#positional-arguments">Positional arguments</a><ul>
<li><a class="reference internal" href="#blast-file">blast_file</a></li>
<li><a class="reference internal" href="#seq-info">seq_info</a></li>
<li><a class="reference internal" href="#taxonomy">taxonomy</a></li>
</ul>
</li>
<li><a class="reference internal" href="#optional-input">Optional input</a><ul>
<li><a class="reference internal" href="#rank-thresholds">rank-thresholds</a></li>
<li><a class="reference internal" href="#copy-numbers">copy-numbers</a></li>
<li><a class="reference internal" href="#weights">weights</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output">Output</a><ul>
<li><a class="reference internal" href="#out">out</a></li>
<li><a class="reference internal" href="#details-out">details-out</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#internal-functions">Internal functions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to the bioy&#8217;s documentation page</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/classifier.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to the bioy’s documentation page"
             >previous</a> |</li>
        <li><a href="index.html">bioy  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012-2014, Noah Hoffman, Chris Rosenthal, Tyler Land.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>